File: backend/app/routers/learning.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.limiter import limiter
from app.dependencies import get_current_user
from app.models.user import User
from app.schemas.learning import (
    GenerateQuizRequest,
    GenerateQuizResponse,
    NextPairRequest,
    NextPairResponse,
    RemediateRequest,
    RemediateResponse,
    StartModuleRequest,
    StartModuleResponse,
    SubmitQuizRequest,
    SubmitQuizResponse,
)
from app.services import learning_service, quiz_service

router = APIRouter(prefix="/learn", tags=["learning"])


@router.post("/start", response_model=StartModuleResponse)
@limiter.limit("5/minute")
async def start_module(
    request: Request,
    data: StartModuleRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    if data.level not in ("kid", "intermediate", "expert"):
        raise HTTPException(status_code=400, detail="Level must be kid, intermediate, or expert")
    return await learning_service.start_module(
        data.topic, data.level, current_user, db, prerequisite_concepts=data.prerequisite_concepts or None
    )


@router.post("/quiz/generate", response_model=GenerateQuizResponse)
@limiter.limit("5/minute")
async def generate_quiz(
    request: Request,
    data: GenerateQuizRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await learning_service.generate_quiz_for_passage(data.passage_id, db)


@router.post("/quiz/submit", response_model=SubmitQuizResponse)
@limiter.limit("5/minute")
async def submit_quiz(
    request: Request,
    data: SubmitQuizRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await quiz_service.score_quiz(data.quiz_id, data.answers, db, data.local_date)
    return SubmitQuizResponse(**result)


@router.post("/passage/next", response_model=NextPairResponse)
@limiter.limit("5/minute")
async def next_passage_pair(
    request: Request,
    data: NextPairRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await learning_service.generate_next_pair(data.module_id, data.covered_concepts, db)


@router.get("/resume/{module_id}", response_model=StartModuleResponse)
@limiter.limit("5/minute")
async def resume_module(
    request: Request,
    module_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await learning_service.resume_module(module_id, db)


@router.post("/remediate", response_model=RemediateResponse)
@limiter.limit("5/minute")
async def remediate(
    request: Request,
    data: RemediateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    if not data.failed_concepts:
        raise HTTPException(status_code=400, detail="No failed concepts provided")
    return await learning_service.remediate(data.module_id, data.quiz_id, data.failed_concepts, db)

File: frontend-web/src/api/learning.ts
import axiosClient from './axiosClient'
import type {
  StartModuleResponse,
  GenerateQuizResponse,
  SubmitQuizResponse,
  NextPairResponse,
  Remediation,
  AnswerSubmission,
} from '../types'

export const startModule = (topic: string, level: string, prerequisite_concepts?: string[]) =>
  axiosClient.post<StartModuleResponse>('/learn/start', { topic, level, prerequisite_concepts: prerequisite_concepts ?? [] })

export const resumeModule = (module_id: string) =>
  axiosClient.get<StartModuleResponse>(`/learn/resume/${module_id}`)

export const generateQuiz = (passage_id: string) =>
  axiosClient.post<GenerateQuizResponse>('/learn/quiz/generate', { passage_id })

export const submitQuiz = (quiz_id: string, answers: AnswerSubmission[]) => {
  const d = new Date()
  const local_date = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`
  return axiosClient.post<SubmitQuizResponse>('/learn/quiz/submit', { quiz_id, answers, local_date })
}

export const nextPair = (module_id: string, covered_concepts: string[]) =>
  axiosClient.post<NextPairResponse>('/learn/passage/next', { module_id, covered_concepts })

export const remediate = (module_id: string, quiz_id: string, failed_concepts: string[]) =>
  axiosClient.post<{ remediations: Remediation[] }>('/learn/remediate', { module_id, quiz_id, failed_concepts })
